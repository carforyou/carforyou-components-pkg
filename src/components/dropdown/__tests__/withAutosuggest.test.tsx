// COPIED form listings-web
// Refactored to @testing-library/react

import React from "react"
import { render, cleanup, fireEvent, within } from "@testing-library/react"

import DropdownWithAutosuggest from "../withAutosuggest"

describe("<DropdownWithAutosuggest>", () => {
  const mockedOnSelect = jest.fn()

  const placeholder = "Select number"
  const options = [
    { value: 1, name: "One" },
    { value: 2, name: "Two" },
    { value: 3, name: "Three" }
  ]

  const mountWrapper = (dropdownOptions = {}) => {
    const { allowCustomValues } = {
      allowCustomValues: false,
      ...dropdownOptions
    }

    return render(
      <DropdownWithAutosuggest
        options={options}
        onSelect={mockedOnSelect}
        input={({ getInputProps }) => (
          <input
            {...getInputProps({
              name: `${name}From`,
              placeholder
            })}
          />
        )}
        allowCustomValues={allowCustomValues}
      />
    )
  }

  const openAutosuggest = wrapper => {
    const { getByPlaceholderText, getByTestId } = wrapper
    const input = getByPlaceholderText(placeholder)

    fireEvent.focus(input)
    getByTestId("dropdown-options")
  }

  afterEach(() => {
    mockedOnSelect.mockClear()
    cleanup()
  })

  describe("when closed", () => {
    it("renders the input", () => {
      const { getByPlaceholderText } = mountWrapper()
      const input = getByPlaceholderText(placeholder)

      expect(input).toMatchSnapshot()
    })
  })

  describe("when opened", () => {
    it("renders all suggestions", () => {
      const wrapper = mountWrapper()
      openAutosuggest(wrapper)

      const { getByTestId } = wrapper
      const list = getByTestId("dropdown-options")
      options.forEach(({ name }) => {
        expect(within(list).getByText(name)).toMatchSnapshot()
      })
    })

    it("marks the partial match when typing", async () => {
      const wrapper = mountWrapper()
      openAutosuggest(wrapper)

      const { getByPlaceholderText, getByTestId } = wrapper
      const input = getByPlaceholderText(placeholder)
      fireEvent.change(input, { target: { value: "On" } })

      const optionsList = getByTestId("dropdown-options")
      const node = within(optionsList).getByTestId("One")
      expect(node).toMatchSnapshot()
    })

    it("highlights elements on exact match when typing", async () => {
      const wrapper = mountWrapper()
      openAutosuggest(wrapper)

      const { getByPlaceholderText, getByTestId } = wrapper
      const input = getByPlaceholderText(placeholder)
      fireEvent.change(input, { target: { value: "One" } })

      const optionsList = getByTestId("dropdown-options")
      const node = within(optionsList).getByTestId("One")
      expect(node).toMatchSnapshot()
    })

    it("select highlited option on blur", async () => {
      const wrapper = mountWrapper()
      openAutosuggest(wrapper)

      const { getByPlaceholderText } = wrapper
      const input = getByPlaceholderText(placeholder)
      fireEvent.change(input, { target: { value: "Two" } })
      fireEvent.blur(input)

      expect(mockedOnSelect).toBeCalledWith(2)
    })

    it("forwards selected value to callback", async () => {
      const wrapper = mountWrapper()
      openAutosuggest(wrapper)

      const { getByText } = wrapper
      const optionNode = getByText("Two")
      fireEvent.click(optionNode)

      expect(mockedOnSelect).toBeCalledWith(2)
    })

    it("clears previous selection on empty input", async () => {
      const wrapper = mountWrapper()
      openAutosuggest(wrapper)

      const { getByPlaceholderText } = wrapper
      const input = getByPlaceholderText(placeholder)

      fireEvent.change(input, { target: { value: "Two" } })
      fireEvent.blur(input)
      fireEvent.change(input, { target: { value: "" } })
      fireEvent.blur(input)

      expect(mockedOnSelect).toBeCalledWith(null)
    })

    it("selects on enter", async () => {
      const wrapper = mountWrapper()
      openAutosuggest(wrapper)

      const { getByPlaceholderText } = wrapper
      const input = getByPlaceholderText(placeholder)

      fireEvent.change(input, { target: { value: "Two" } })
      fireEvent.keyDown(input, {
        keyCode: 13,
        target: { blur: () => fireEvent.blur(input) }
      })

      expect(mockedOnSelect).toBeCalledWith(2)
    })
  })

  describe("with custom values", () => {
    it("allows custom user input", async () => {
      const { getByPlaceholderText } = mountWrapper({ allowCustomValues: true })
      const input = getByPlaceholderText(placeholder)
      fireEvent.change(input, { target: { value: "Custom" } })
      fireEvent.blur(input)

      expect(mockedOnSelect).toBeCalledWith({ customValue: "Custom" })
    })

    it("allows option from dropdown", async () => {
      const { getByPlaceholderText } = mountWrapper({ allowCustomValues: true })
      const input = getByPlaceholderText(placeholder)
      fireEvent.change(input, { target: { value: "Two" } })
      fireEvent.blur(input)

      expect(mockedOnSelect).toBeCalledWith(2)
    })
  })
})
