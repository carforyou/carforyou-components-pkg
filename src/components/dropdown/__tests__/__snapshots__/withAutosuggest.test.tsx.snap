// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownWithAutosuggest> when closed renders the input 1`] = `
<input
  aria-autocomplete="list"
  aria-labelledby=""
  autocomplete="none"
  autocorrect="off"
  class="cursor-text w-12/12 overflow-hidden"
  data-testid="From"
  id="From"
  name="From"
  placeholder="Select number"
  spellcheck="false"
  value=""
/>
`;

exports[`<DropdownWithAutosuggest> when opened highlights elements on exact match when typing 1`] = `
<li
  aria-selected="false"
  class="hover:bg-grey-bright transition duration-200 cursor-pointer px-20 py-10"
  data-testid="One"
  id="downshift-3-item-0"
  role="option"
>
  
  <span
    class="font-bold underline"
  >
    One
  </span>
  
</li>
`;

exports[`<DropdownWithAutosuggest> when opened marks the partial match when typing 1`] = `
<li
  aria-selected="false"
  class="hover:bg-grey-bright transition duration-200 cursor-pointer px-20 py-10"
  data-testid="One"
  id="downshift-2-item-0"
  role="option"
>
  
  <span
    class="font-bold underline"
  >
    On
  </span>
  e
</li>
`;

exports[`<DropdownWithAutosuggest> when opened renders all suggestions 1`] = `
<li
  aria-selected="false"
  class="hover:bg-grey-bright transition duration-200 cursor-pointer px-20 py-10"
  data-testid="One"
  id="downshift-1-item-0"
  role="option"
>
  One
</li>
`;

exports[`<DropdownWithAutosuggest> when opened renders all suggestions 2`] = `
<li
  aria-selected="false"
  class="hover:bg-grey-bright transition duration-200 cursor-pointer px-20 py-10"
  data-testid="Two"
  id="downshift-1-item-1"
  role="option"
>
  Two
</li>
`;

exports[`<DropdownWithAutosuggest> when opened renders all suggestions 3`] = `
<li
  aria-selected="false"
  class="hover:bg-grey-bright transition duration-200 cursor-pointer px-20 py-10"
  data-testid="Three"
  id="downshift-1-item-2"
  role="option"
>
  Three
</li>
`;
